src/com/amazon/corretto/crypto/provider/Loader.java

```diff
@@ -289,68 +342,116 @@ private static void assertVersionMatch() {
-                    String hexByte = Integer.toHexString(Byte.toUnsignedInt(b));
-                    if (hexByte.length() == 1) {
-                        fileName.append('0');
-                    }
-                    fileName.append(hexByte);
-                }
-                fileName.append(suffix);
-
-                final Path tmpFile = tmpDir.resolve(fileName.toString());
-
-                try {
-                    final Path result = Files.createFile(tmpFile, permissions);
-                    if (DebugFlag.VERBOSELOGS.isEnabled()) {
-                        LOG.log(Level.FINE, "Created temporary library file after " + attempt + " attempts");
-                    }
-                    result.toFile().deleteOnExit();
-                    return result;
-                } catch (final FileAlreadyExistsException ex) {
-                    // We ignore and retry this exception
-                } catch (final Exception ex) {
-                    // Any other exception is bad and we may need to quash.
-                    throw new AssertionError("Unable to create temporary file");
+    /**
+     * Unfortunately, we cannot actually use Files.createTempFile, because that internally depends on
+     * SecureRandom, which results in a circular dependency.
+     */
+    private static synchronized Path createPrivateTmpDir(final String prefix) throws IOException {
+        final Path systemTempDir = Paths.get(System.getProperty("java.io.tmpdir"));
+        if (!Files.isDirectory(systemTempDir)) {
+            throw new AssertionError("java.io.tmpdir is not valid: " + systemTempDir);
+        }
+
+        final byte[] rndBytes = bootstrapRng(Long.BYTES); // Default java tmp files use this much entropy
+        final StringBuilder privateTempDir = new StringBuilder(prefix);
+
+        for (byte b : rndBytes) {
+            // We convert to an unsigned integer first to avoid sign-bit extension when converting to hex.
+            String hexByte = Integer.toHexString(Byte.toUnsignedInt(b));
+            if (hexByte.length() == 1) {
+                privateTempDir.append('0');
+            }
+            privateTempDir.append(hexByte);
+        }
+
+        final Path privateDirFullPath = systemTempDir.resolve(privateTempDir.toString());
+        final int RETRY_LIMIT = 10;
+        int attempt = 0;
+
+        while(attempt < RETRY_LIMIT) {
+            attempt++;
+            try {
+                final Path result = Files.createDirectory(privateDirFullPath, SELF_OWNER_FILE_PERMISSIONS);
+                if (DebugFlag.VERBOSELOGS.isEnabled()) {
+                    LOG.log(Level.FINE, "Created temporary library directory");
                 }
+                result.toFile().deleteOnExit();
+                return result;
+            } catch (final FileAlreadyExistsException ex) {
+                // We ignore and retry this exception
```

Comment:
why retry this without taking any action? if the file already exists, we'd need to delete it before we'd expect our directory creation to work, right?

Reply:
Random generation *should* be in the retry loop so it goes for different names. If not then that's a bug.

Reply:
Nice catch! I don't think we should delete the directory since if it already exists then it was created by something else and may be a collision. We already call `result.toFile().deleteOnExit();` so we will automatically delete our own temp directory. 

I've moved the call to `bootstrapRng()` to be inside the retry loop so that a different directory name is generated on each retry.